### PR TITLE
Fix #2772 Load Google Fonts async

### LINKED ISSUE #2772: Load Google Fonts asynchronously
**Describe the solution you'd like**
We are missing one enhancement to our Google Fonts optimization feature.

Like it's described on this post, we could load asynchronously the Google Fonts:
https://csswizardry.com/2020/05/the-fastest-google-fonts/

This method is also going to be implemented to load the CSS files when the « Optimize CSS delivery » option is activated:
https://github.com/wp-media/wp-rocket/issues/2658

### PR DESCRIPTION
**Pull request by crystinutzaa**
#### inc/Engine/Optimization/GoogleFonts/Combine.php
- Rename the class to `Optimize`, to have it make more sense about what it is doing now
- Add a class alias in `deprecated/3.6.php` for the old name
- Add a new method `preload()` to add a preload link on the page for the Google Fonts link
- Add a new method `async()` to use the `media="print"` solution for asynchronous, non-blocking fetching of the link
- Add a new method `fallback()` to insert a `noscript` fallback in the footer
- Use the 3 above methods in the `optimize()` method to apply everything needed

We'll also have to make sure we don't double apply those optimization when `Optimize CSS Delivery` is enabled at the same time as `Optimize Google Fonts`.

### Estimate the effort ✅ 
Effort `[M]`

TO DO 
- [x] - Revisit Unit tests for Async CSS
- [x] - Revisit Integration test  Async CSS
- [x] - Revisit Unit tests for Google Fonts changes to reflect the changes with Async
- [x] - Revisit Integration tests for Google fonts changes to reflect the changes with Async

Blocking this one due to [DOM Module](https://github.com/wp-media/module-dom)
Exactly as in PR https://github.com/wp-media/wp-rocket/pull/2786 , we need DomDocument to be able to handle easily the link attributes.

### CODE REVIEW

**Comment by crystinutzaa** [comment]
Blocking this one due to [DOM Module](https://github.com/wp-media/module-dom)
Exactly as in PR https://github.com/wp-media/wp-rocket/pull/2786 , we need DomDocument to be able to handle easily the link attributes.

**Comment by hellofromtonya** [comment]
The async css part of this PR is already done in PR 2786 as any `<link rel="stylesheet"...>` gets the async css treatment. You can see it here in the [Async CSS test data](https://github.com/wp-media/wp-rocket/blob/master/tests/Fixtures/inc/Engine/CriticalPath/content/twentyseventeen-async_css.html#L37-L38) where the Google fonts `<link>`:

original:

```html
<link rel="stylesheet" id="twentyseventeen-fonts-css" href="https://fonts.googleapis.com/css?family=Libre+Franklin%3A300%2C300i%2C400%2C400i%2C600%2C600i%2C800%2C800i&amp;subset=latin%2Clatin-ext&amp;display=fallback" media="all">
```

after Async CSS runs:

```html
<link rel="preload" href="https://fonts.googleapis.com/css?family=Libre+Franklin%3A300%2C300i%2C400%2C400i%2C600%2C600i%2C800%2C800i&amp;subset=latin%2Clatin-ext&amp;display=fallback" as="style">
<link rel="stylesheet" id="twentyseventeen-fonts-css" href="https://fonts.googleapis.com/css?family=Libre+Franklin%3A300%2C300i%2C400%2C400i%2C600%2C600i%2C800%2C800i&amp;subset=latin%2Clatin-ext&amp;display=fallback" media="print" onload="this.onload=null;this.media='all'">
```

**Comment by Tabrisrp** [comment]
I believe this PR is to do this when the `Optimize CSS Delivery` option is not enabled, but the `Optimize Google Fonts` is

**Comment by hellofromtonya** [comment]
>I believe this PR is to do this when the Optimize CSS Delivery option is not enabled, but the Optimize Google Fonts is

Then let's not duplicate code here. Instead, refactor to reuse. How? 

- Pull the `AsyncCSS` class out of `CriticalPath` module. 
- Add a specific implementation for the query and rules into each module.

**Review by remyperona — CHANGES_REQUESTED** [review]

File: inc/Engine/CriticalPath/CriticalCSSSubscriber.php [hunk 1]
```diff
@@ -583,14 +584,48 @@ public function exclude_inline_js( array $excluded_inline ) {
...
+	/**
+	 * Checks if we should apply deferring of CSS files.
+	 *
+	 * @return bool True if we should, false otherwise.
+	 */
+	private function maybe_async_css() {
+		if (
+			rocket_get_constant( 'DONOTROCKETOPTIMIZE' )
+			||
+			rocket_get_constant( 'DONOTASYNCCSS' )
```
[the remark this answers is not in the document]
Comment (crystinutzaa) [reply inc/Engine/CriticalPath/CriticalCSSSubscriber.php:609]: Which one: `DONOTASYNCCSS` ?

[the remark this answers is not in the document]
Comment (Tabrisrp) [reply inc/Engine/CriticalPath/CriticalCSSSubscriber.php:609]: Yes

**Review by remyperona — APPROVED** [review]

**Comment by GeekPress** [comment]
I noticed that this enhancement relay on « Optimize CSS Delivery » option turned on.

Why don't we directly apply this enhancement? What is the relation with our CPCSS? I suppose that even if we use CPCSS, we will still have a small FOUC on fonts as the CSS doesn't depend to the internal CSS.

**Comment by GeekPress** [comment]
@arunbasillal What are thoughts about that? 
https://github.com/wp-media/wp-rocket/pull/2834#issuecomment-659443323

**Comment by hellofromtonya** [comment]
>I noticed that this enhancement relay on « Optimize CSS Delivery » option turned on.
>
>Why don't we directly apply this enhancement? What is the relation with our CPCSS? I suppose that even if we use CPCSS, we will still have a small FOUC on fonts as the CSS doesn't depend to the internal CSS.

@GeekPress Where are you noticing this?

Looking at this PR, each option separately handles how async css is applied.

How? We moved the `AsyncCSS` out of the Critical Path module and are now using for any async css needs. Both the Optimize CSS Delivery and Google Font Optimization use it separately. In other words, Google Font Optimization is not dependent upon OCD and OCD is not dependent upon Google Font Optimization. They are separated.

**Comment by GeekPress** [comment]
@hellofromtonya Don't this check if the Optimize CSS Delivery is on?
https://github.com/wp-media/wp-rocket/pull/2834/files#diff-5ac8e03c0609dc96a4285b6984b74d66R612

**Comment by hellofromtonya** [comment]
> Don't this check if the Optimize CSS Delivery is on?

Yes for the Optimize CSS Delivery only, i.e. not Google Fonts Optimization.

**Comment by GeekPress** [comment]
> Yes for the Optimize CSS Delivery only, i.e. not Google Fonts Optimization.

Ok, thanks for the confirmation 😇

**Pull request closed without merging**